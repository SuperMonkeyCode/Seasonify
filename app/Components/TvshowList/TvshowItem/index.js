/* @flow */
import { connect } from 'react-redux';

import { tvshowSelectors } from 'Store/tvshowStore';
import { viewingSelectors } from 'Store/viewingStore';

import TvshowItem from './TvshowItem';

export default connect(
  () => (state, { tvshowId }) => ({
    seasonsCount: tvshowSelectors.makeGetSeasonsCount()(state, { tvshowId }),
    tvshow: tvshowSelectors.getTvshow(state, { tvshowId }),
    viewers: viewingSelectors.makeGetViewersArray()(state, { tvshowId }),
  }),
  {},
)(TvshowItem);
