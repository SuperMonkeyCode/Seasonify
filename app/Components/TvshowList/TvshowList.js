/* @flow */
import { Container, Content } from 'native-base';
import { /* FlatList,*/ VirtualizedList } from 'react-native';
import {
  branch,
  compose,
  hoistStatics,
  pure,
  renderComponent,
  withHandlers,
} from 'recompose';
import React from 'react';

import Loading from 'Components/Loading';
import SingleFAB from 'Components/SingleFAB';
import TvshowItem from './TvshowItem';

type Props = {
  /* parent */
  navigation: Object,
  /* connect */
  tvshowsIds: string[],
  isLoading: LoadingValue,
  /* HOC */
  handleFAB: Function,
};

const enhance = compose(
  pure,
  withHandlers({
    handleFAB: ({ navigation }: Props) => () => {
      navigation.navigate('TvshowAddPage');
    },
  }),
  branch(({ isLoading }: Props) => isLoading, renderComponent(Loading)),
);

function TvshowList({ navigation, tvshowsIds, handleFAB }: Props) {
  /* istanbul ignore next */
  const renderItem = ({ item: id }) =>
    <TvshowItem tvshowId={id} navigate={navigation.navigate} />;

  const getItem = (data, index) => data[index];
  const keyExtractor = item => item;

  return (
    <Container>
      <Content>
        <VirtualizedList
          initialNumberToRender={11}
          data={tvshowsIds}
          getItemCount={/* istanbul ignore next */ data => data.length}
          getItem={getItem}
          renderItem={renderItem}
          keyExtractor={keyExtractor}
        />
      </Content>
      <SingleFAB icon="add" onPress={handleFAB} />
    </Container>
  );
}

export default enhance(TvshowList);
