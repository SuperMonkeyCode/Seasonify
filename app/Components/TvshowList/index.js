/* @flow */
import { connect } from 'react-redux';
import React from 'react';

import { tvshowSelectors } from 'State/tvshowState';
import HeaderRoot from 'Components/HeaderRoot';

import TvshowList from './TvshowList';

TvshowList.navigationOptions = ({ navigation }) => ({
  header: <HeaderRoot title="TV Shows" navigation={navigation} />,
});

export default connect(
  state => ({
    tvshows: tvshowSelectors.getTvshows(state),
  }),
  {},
)(TvshowList);
