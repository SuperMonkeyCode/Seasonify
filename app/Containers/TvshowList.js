/* @flow */
import { Container, Content } from 'native-base';
import { /* FlatList,*/ VirtualizedList } from 'react-native';
import { connect } from 'react-redux';
import React from 'react';

import type { Tvshows } from 'Types';
import HeaderRoot from 'Components/HeaderRoot';
import SingleFAB from 'Components/SingleFAB';
import TvshowItem from 'Components/TvshowItem';
import tvshowState from 'State/tvshowState';

const mapStateToProps = state => ({
  tvshows: tvshowState.selectors.getTvshows(state),
});

const mapActionsToProps = {};

export function TvshowList({ navigation, tvshows }: { navigation: Object, tvshows: Tvshows }) {
  const onFAB = () => {
    navigation.navigate('TvshowAddPage');
  };

  /* istanbul ignore next */
  const renderItem = ({ item: tvshow }) =>
    (<TvshowItem
      tvshowId={tvshow.id}
      tvshowName={tvshow.name}
      navigate={navigation.navigate}
    />);

  const keyExtractor = item => item.id;

  const tvshowKeys = Object.keys(tvshows);

  const getItem = (data, index) => data[tvshowKeys[index]];

  return (
    <Container>
      <Content>
        <VirtualizedList
          initialNumberToRender={11}
          data={tvshows}
          extraData={tvshowKeys}
          getItemCount={/* istanbul ignore next */ data => Object.keys(data).length}
          getItem={getItem}
          renderItem={renderItem}
          keyExtractor={keyExtractor}
        />
      </Content>
      <SingleFAB icon="add" onPress={onFAB} />
    </Container>
  );
}

TvshowList.navigationOptions = ({ navigation }) => ({
  header: <HeaderRoot title="TV Shows" navigation={navigation} />,
});

export default connect(mapStateToProps, mapActionsToProps)(TvshowList);
