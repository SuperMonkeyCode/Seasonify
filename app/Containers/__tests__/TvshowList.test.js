import React from 'react';
import { TvshowList } from '../TvshowList';

function setup(specificProps = {}) {
  const props = {
    tvshows: {},
    navigation: { navigate: jest.fn() },
    ...specificProps,
  };
  const component = shallow(<TvshowList {...props} />);
  return {
    component,
    props,
    actions: {},
  };
}

describe('rendering', () => {
  it('should show 2 series', () => {
    const { component } = setup({
      tvshows: {
        abc: { id: 'abc', name: 'tvshow1' },
        qsd: { id: 'qsd', name: 'tvshow2' },
      },
    });
    expect(component.find('TvshowItem')).toHaveLength(2);
    expect(component).toMatchSnapshot();
  });

  describe('when clicking on FAB', () => {
    it('should navigate', () => {
      const { component, props } = setup();
      component.find('SingleFAB').simulate('press');
      expect(props.navigation.navigate).toBeCalledWith('TvshowAddPage');
    });
  });
});
