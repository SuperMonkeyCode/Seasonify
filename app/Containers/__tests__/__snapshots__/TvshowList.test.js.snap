// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should show 2 series 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Styled(List)>
      <TvshowItem
        navigate={[Function]}
        tvshowId="abc"
        tvshowName="tvshow1"
      />
      <TvshowItem
        navigate={[Function]}
        tvshowId="qsd"
        tvshowName="tvshow2"
      />
    </Styled(List)>
  </Styled(Content)>
  <SingleFAB
    icon="add"
    onPress={[Function]}
  />
</Styled(Container)>
`;
